airflow/jobs/scheduler_job_runner.py

```diff
@@ -1838,12 +1890,61 @@ def _orphan_unreferenced_datasets(self, session: Session = NEW_SESSION) -> None:
     def _executor_to_tis(self, tis: list[TaskInstance]) -> dict[BaseExecutor, list[TaskInstance]]:
         """Organize TIs into lists per their respective executor."""
         _executor_to_tis: defaultdict[BaseExecutor, list[TaskInstance]] = defaultdict(list)
-        executor: str | None
         for ti in tis:
-            if ti.executor:
-                executor = str(ti.executor)
+            if executor_obj := self._try_to_load_executor(ti.executor):
+                _executor_to_tis[executor_obj].append(ti)
             else:
-                executor = None
-            _executor_to_tis[ExecutorLoader.load_executor(executor)].append(ti)
+                continue
 
         return _executor_to_tis
+
+    def _slots_free_for_tis(self, tis: list[TaskInstance]) -> set[TaskInstance]:
+        """Return TIs that we have slots available for."""
+        # First get a mapping of executor names to slots they have available
+        executor_to_slots_available: dict[ExecutorName, int] = {}
+        for executor in self.job.executors:
+            if TYPE_CHECKING:
+                # All executors should have a name if they are initted from the executor_loader. But we need
+                # to check for None to make mypy happy.
+                assert executor.name
+            executor_to_slots_available[executor.name] = executor.slots_available
+
+        # Loop through all the TIs we're scheduling for and add them to a set to be moved to queued if the
+        # executor they're assigned to has slots available.
+        tis_we_have_room_for = set()
+        for ti in tis:
+            if executor_obj := self._try_to_load_executor(ti.executor):
+                ti_exec_name = executor_obj.name
+            else:
+                continue
+
+            if ti_exec_name:
+                if executor_to_slots_available[ti_exec_name] > 0:
+                    tis_we_have_room_for.add(ti)
+                    executor_to_slots_available[ti_exec_name] -= 1
```

Comment:
this, essentially
```suggestion
for ti in tis:
    if executor_obj := self._try_to_load_executor(ti.executor):
        if executor_to_slots_available[executor_obj.name] > 0:
            tis_we_have_room_for.add(ti)
            executor_to_slots_available[executor_obj.name] -= 1
    else:
        continue
```

Reply:
You still need the if no matter where you move the block to make mypy happy (so it can be sure it's not None type), I can nest it where you're asking bu then you have three levels of ifs which looks a bit gross imho, but if it satisfies others I'm happy to do it.

Reply:
Moved the statement up, still with the if, hopefully it's at least a bit more pleasing this way.